OIDC Provider Refactor

NickMeves opened the pull request:
This refactors the OIDC Provider to be much cleaner. It additionally opens up the OIDC logic to be global and be consumed by any OIDC compliant providers (Keycloak, Azure & Gitlab have come up in other PRs).

## Description

- Refactor the ProfileURL fallback logic to `EnrichSession`. This additionally pulls `groups` data if present as well as `email`. This is still only called if claims extraction doesn't fully populate fields.
- Make all OIDC IDToken/Claims to Session logic global to all providers and available on the `ProviderData` object.
- Deprecate referring to Email as `UserID` in code. `oidc-email-claim` is a 1:1 replacement for `user-id-claim`. `user-id-claim` still works for backwards compatibility. Why? - Custom claims to session mapping should be handled in our future goal of adhoc claims.
- LOTS of new tests for areas that were untested.
  - I want to cleanup existing tests and potentially move to Ginkgo. But I will do that in another PR so the existing test suite still runs and validates the refactor.

## Motivation and Context

OIDC is our main provider but it has gotten a lot of code smell as a bunch of contributors have added improvements adhoc. This tries to add some strategy to the design of the provider.

Base-OIDC made global is based on working with PRs for other providers that seem to be wanting OIDC-like functionality. Making it global will also make multiple providers ready for adhoc IDToken Claim -> Session -> Header/Userinfo that former Louketo users have been asking for.

## How Has This Been Tested?

Added unit tests to any new functionality & to migrated methods that were previously untested.

Left ugly tests as-is for now in the OIDC provider to ensure the existing test suite still passes during the refactor. Will rewrite them to test tables or ginkgo in another PR.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.
- [ ] I have tested interactively with an OIDC provider

NickMeves commented on providers/oidc.go:
```diff
@@ -52,23 +42,74 @@ func (p *OIDCProvider) Redeem(ctx context.Context, redirectURL, code string) (s
+		s.Email = email
+	}
+
+	// Handle array & singleton groups cases
```
This feels ugly... Have a better idea for singtleton + list support from JSON methods?

JoelSpeed replied:
Someone has to coerce it to a list, if the json doesn't do that automagically we have to. Might be nice as a helper though?

NickMeves replied:
Yeah, I'll whip up a helper for this. That feels cleaner and will reduce the conditional complexity of this method.

NickMeves commented on providers/oidc.go:
```diff
@@ -109,19 +152,14 @@ func (p *OIDCProvider) redeemRefreshToken(ctx context.Context, s *sessions.Sessi
 		return fmt.Errorf("failed to get token: %v", err)
 	}
 
-	// in the token refresh response the id_token is optional
```
The refresh vs redeem verified IDToken requirements were moved into `createSession` to reduce redundant code and logic branches.

NickMeves commented on providers/oidc.go:
```diff
@@ -135,193 +173,59 @@ func (p *OIDCProvider) redeemRefreshToken(ctx context.Context, s *sessions.Sessi
-	return nil, nil
-}
-
-func (p *OIDCProvider) createSessionState(ctx context.Context, token *oauth2.Token, idToken *oidc.IDToken) (*sessions.SessionState, error) {
```
This is replaced by `createSession` and is still OIDCProvider specific in this file and not moved with other global OIDC helper methods.

NickMeves commented on providers/oidc.go:
```diff
@@ -135,193 +173,59 @@ func (p *OIDCProvider) redeemRefreshToken(ctx context.Context, s *sessions.Sessi
-
-	newSession.Email = claims.UserID // TODO Rename SessionState.Email to .UserID in the near future
-
-	newSession.User = claims.Subject
```
They assignments were redundant - they were already done in `buildSessionFromClaims`

NickMeves commented on providers/provider_data.go:
```diff
@@ -94,3 +110,99 @@ func defaultURL(u *url.URL, d *url.URL) *url.URL {
+	ss.Email = claims.Email
+	ss.Groups = claims.Groups
+
+	// TODO (@NickMeves) Deprecate for dynamic claim to session mapping
```
Stop treating `PreferredUsername` as a top-level citizen and get it ready to be a generic optional claim that is extracted from raw claims.

NickMeves commented on providers/provider_data.go:
```diff
@@ -94,3 +110,99 @@ func defaultURL(u *url.URL, d *url.URL) *url.URL {
+		ss.PreferredUsername = pref
+	}
+
+	// `email_verified` must be present and explicitly set to `false` to be
```
I'm pretty sure this is the intention of this logic? It tripped me up in writing new unit tests when I made a case with no `email_verified` claim.

I'm assuming we want those to pass even if `AlloUnverifiedEmails` is false? We only want to explicitly deny `"email_verified": false`?

NickMeves replied:
(Only the comment is new, the rest is copy/paste from the other file).

JoelSpeed replied:
I think that logic is correct yes 🤔

NickMeves commented on providers/provider_data_test.go:
```diff
@@ -0,0 +1,414 @@
+	}
+}
+
+func TestProviderData_verifyIDToken(t *testing.T) {
```
Ok, I lied. This is the one legacy unit test of a private method that I ripped out and rewrote to use a test table. In my defense, I also restructured the private method itself to use the new `getIDToken` helper to reduce manual `id_token` extraction from `oauth2.Token` that was in a few spots.

NickMeves commented:
Mmmmm - satisfying 😃  https://codeclimate.com/github/oauth2-proxy/oauth2-proxy/pull/936

JoelSpeed commented on providers/oidc.go:
```diff
@@ -238,37 +240,41 @@ func (p *OIDCProvider) findVerifiedIDToken(ctx context.Context, token *oauth2.To
+	// TODO (@NickMeves) Deprecate for dynamic claim to session mapping
+	if pref, ok := claims.rawClaims["preferred_username"].(string); ok {
+		ss.PreferredUsername = pref
+	}
```
This reminds me of one thought I had before, I wonder if we could just use the easyjson library to extract the claims 🤔 We could come up with some helper for all available claims and set them appropriately using some mapping

JoelSpeed commented on providers/oidc.go:
```diff
@@ -68,7 +68,7 @@ func (p *OIDCProvider) EnrichSession(ctx context.Context, s *sessions.SessionSta
 
 	// Try to get missing emails or groups from a profileURL
-	if s.Email == "" || s.Groups == nil {
+	if s.Email == "" || len(s.Groups) == 0 {
```
What if they are a member of 0 groups, does this not duplicate the work? Is there a way we can tell if the group check has already been done?

NickMeves replied:
I think the scenario comes up for users who don't set a `scope` that will appropriately populate the claims with a groups/role claim.  But their OIDC IdP Userinfo endpoint populates the data (I'm assuming this is the use case).

Here's the link to the discussion: https://github.com/oauth2-proxy/oauth2-proxy/issues/883#issuecomment-723972563

I wonder if I can figure out something crafty between `nil` vs `[]string{}` to know if a groups claim was present but empty?

I had the same thoughts about extra HTTP requests during the redeem flow -- I couldn't figure out a way to know on our end to know for sure the userinfo endpoint wouldn't have groups data if the IDToken claims was missing it 🤔

NickMeves replied:
Here's my shot at resolving this in this commit: https://github.com/oauth2-proxy/oauth2-proxy/pull/936/commits/154af7acae1e927f4dc3b71ef00428916a97d8ba

JoelSpeed commented on providers/oidc_test.go:
```diff
@@ -233,6 +233,256 @@ func TestOIDCProviderRedeem_custom_userid(t *testing.T) {
+				RefreshToken: refreshToken,
+			},
+		},
+
```
Nit random empty line

JoelSpeed commented on pkg/validation/options.go:
```diff
@@ -233,7 +233,10 @@ func parseProviderInfo(o *options.Options, msgs []string) []string {
+	// Make the OIDC options available to all providers that support it
+	p.AllowUnverifiedEmail = o.InsecureOIDCAllowUnverifiedEmail
+	p.EmailClaim = o.UserIDClaim
+	p.GroupsClaim = o.OIDCGroupsClaim
```
Would be really nice to have some sort of provider constructor like we do in the session storage so we don't have to do this here 🤔

NickMeves replied:
Good idea. I can whip something up to help with this.

One of my next provider refactors I want to take on is getting a `Configure` method or something on each provider so the options they consume is much more clear in general.

NickMeves replied:
How do you feel about the placement of the globally available OIDC methods? `ProviderData` felt most global, but I'm not in love with it since the OIDC methods are a bit different than `ProviderData` type fields.

I also noticed both `internal_util` and `util` in that package, but I'll leave that consolidation for another PR.

JoelSpeed commented on pkg/apis/options/options.go:
```diff
@@ -241,7 +244,7 @@ func NewFlagSet() *pflag.FlagSet {
 	flagSet.Bool("gcp-healthchecks", false, "Enable GCP/GKE healthcheck endpoints")
 
-	flagSet.String("user-id-claim", "email", "which claim contains the user ID")
+	flagSet.String("user-id-claim", providers.OIDCEmailClaim, "(DEPRECATED) which claim contains the user ID")
```
Should note which option replaced it here IMO

JoelSpeed commented on CHANGELOG.md:
```diff
@@ -43,6 +43,7 @@
 
 ## Changes since v6.1.1
 
+- [#936](https://github.com/oauth2-proxy/oauth2-proxy/pull/936) Refactor OIDC Provider and support groups from Profile URL (@NickMeves)
```
Need to add a note to the changelog about the deprecated option

JoelSpeed reviewed (commented):
Ok, I went through this commit by commit, turns out that wasn't a great idea but hey ho 😅 I've added a bunch of comments but GH might mark some as resolved as I think some code got moved after I added the comments, PTAL below instead of in the code view

NickMeves commented on providers/oidc.go:
```diff
@@ -180,51 +182,51 @@ func (p *OIDCProvider) CreateSessionFromToken(ctx context.Context, token string)
+	// Allow empty Email in Bearer case since we can't hit the ProfileURL
+	if ss.Email == "" {
+		ss.Email = ss.User
+	}
```
This was always the case. When an email isn't present, the `sub` is used as the email and users have to use `--email-domains` `*` to make it work.

You can see the parallel in the generic bearer token to session logic here: https://github.com/oauth2-proxy/oauth2-proxy/blob/f6ae15e8c37b15c7cc29332c1f070a06bc503dc7/pkg/apis/middleware/session.go#L38

Ideally, when we sort out this PR (https://github.com/oauth2-proxy/oauth2-proxy/pull/770) we can get rid of this hackiness that is all over the code to support providers that don't always have an email.

NickMeves commented on providers/oidc.go:
```diff
@@ -52,23 +52,64 @@ func (p *OIDCProvider) Redeem(ctx context.Context, redirectURL, code string) (s
+	return nil
+}
+
+// ValidateSessionState checks that the session's IDToken is still valid
```
Darn, it looks like my IDE did a bad job refactoring all instances of method names in the naming refactor PR and missed some comment, I'll try to handle them here.

NickMeves replied:
https://github.com/oauth2-proxy/oauth2-proxy/pull/938

NickMeves commented:
> Ok, I went through this commit by commit, turns out that wasn't a great idea but hey ho 😅 I've added a bunch of comments but GH might mark some as resolved as I think some code got moved after I added the comments, PTAL below instead of in the code view

You got to see my stream of consciousness waffling on the right place to organize the central provider agnostic OIDC help methods 😅 

And after all that, I'm still not sure I like where I ended up 🤷‍♂️

JoelSpeed commented on CHANGELOG.md:
```diff
@@ -43,6 +44,8 @@
 
 ## Changes since v6.1.1
 
+- [#816](https://github.com/oauth2-proxy/oauth2-proxy/pull/816) (via [#936](https://github.com/oauth2-proxy/oauth2-proxy/pull/936)) Support non-list group claims (@loafoe)
```
Never had a via before, works though!

NickMeves replied:
I wanted to give credit for discovering the issue and implementing a potential solution -- but I also didn't want to merge that PR then to immediately rip it out and refactor the concept somewhere else. This felt like good compromise.

NickMeves commented:
Commit for the coerce list note in simpleJSON profileURL groups extraction: https://github.com/oauth2-proxy/oauth2-proxy/pull/936/commits/61d43b481fdeb7e90c86fe9bb0454bfb644c7683

Scope creeped it a bit, added support for complex group structures in the profileURL too (instead of just IDToken claims).

JoelSpeed commented on providers/util.go:
```diff
@@ -55,3 +59,42 @@ func makeLoginURL(p *ProviderData, redirectURI, state string, extraParams url.Va
+// formatGroup coerces an OIDC groups claim into a string
+// If it is non-string, marshal it into JSON.
+func formatGroup(rawGroup interface{}) (string, error) {
+	group, ok := rawGroup.(string)
+	if !ok {
+		jsonGroup, err := json.Marshal(rawGroup)
+		if err != nil {
+			return "", err
+		}
+		group = string(jsonGroup)
+	}
+	return group, nil
```
How about reformatting this to remove some of the indentation?
```suggestion
	if group, ok := rawGroup.(string); {
	  return group, nil
	}
	
	jsonGroup, err := json.Marshal(rawGroup)
	if err != nil {
		return "", err
	}
	return string(jsonGroup), nil
```

JoelSpeed commented on providers/oidc.go:
```diff
@@ -52,23 +43,73 @@ func (p *OIDCProvider) Redeem(ctx context.Context, redirectURL, code string) (s
-	s, err = p.createSessionState(ctx, token, idToken)
+// callProfileURL enriches a session's Email & Groups via the JSON response of
+// an OIDC profile URL
+func (p *OIDCProvider) callProfileURL(ctx context.Context, s *sessions.SessionState) error {
```
Should the name of this method be a bit more descriptive? Perhaps `enrichFromProfileURL`?

JoelSpeed commented on providers/provider_data.go:
```diff
@@ -94,3 +110,116 @@ func defaultURL(u *url.URL, d *url.URL) *url.URL {
+
+func (p *ProviderData) verifyIDToken(ctx context.Context, token *oauth2.Token) (*oidc.IDToken, error) {
+	rawIDToken := getIDToken(token)
+	if strings.TrimSpace(rawIDToken) != "" {
+		if p.Verifier == nil {
+			return nil, ErrMissingOIDCVerifier
+		}
+		return p.Verifier.Verify(ctx, rawIDToken)
+	}
+	return nil, ErrMissingIDToken
```
Unindent maybe?
```suggestion
	if strings.TrimSpace(rawIDToken) == "" {
		return nil, ErrMissingIDToken
	}
	if p.Verifier == nil {
		return nil, ErrMissingOIDCVerifier
	}
	return p.Verifier.Verify(ctx, rawIDToken)
```

NickMeves replied:
Good catch! I'm with you on less nested conditionals, I made these changes and a few more I caught when eyeballing this PR again. Thanks.

JoelSpeed reviewed (commented):
Added a few suggestions, looks like it might be in need of a rebase as well? I saw some of the comments still had the wrong names 🤔

JoelSpeed reviewed (dismissed):
LGTM! Thanks @NickMeves

JoelSpeed commented:
Did you want to squash at all before we merge?

NickMeves commented:
@JoelSpeed Everything should be cleaned up and ready to roll.  Thanks for the review!
